src/pressure/VCycleMG.cpp

```diff
@@ -753,154 +612,244 @@ void VCycleMG::Solve(Field &out, Field &tmp, Field const &b, size_t level, bool
+}
+void VCycleMG::call_solve_colored_gauss_seidel(Field &out, Field &tmp, Field const &b, const size_t level, bool sync) {
+    auto domain = Domain::getInstance();
+
+    // local variables and parameters for GPU
+    const size_t Nx = domain->get_Nx(level);
+    const size_t Ny = domain->get_Ny(level);
+
+    const real dx = domain->get_dx(level);
+    const real dy = domain->get_dy(level);
+    const real dz = domain->get_dz(level);
+
+    real *data_out = out.data;
+    FieldType type = out.get_type();
+
+    BoundaryController *boundary = BoundaryController::getInstance();
+
+    size_t *data_inner_list = boundary->get_inner_list_level_joined();
+    const size_t bsize_i = boundary->get_size_inner_list_level_joined();
+
+    // start/end
     // inner
-    size_t start_i = boundary->get_inner_list_level_joined_start(level);
-    size_t end_i = boundary->get_inner_list_level_joined_end(level) + 1;
-    size_t start_b = boundary->get_boundary_list_level_joined_start(level);
-    size_t end_b = boundary->get_boundary_list_level_joined_end(level) + 1;
+    const size_t start_i = boundary->get_inner_list_level_joined_start(level);
+    const size_t end_i = boundary->get_inner_list_level_joined_end(level) + 1;
 
-    boundary->apply_boundary(out.data, level, type, sync);
+    boundary->apply_boundary(data_out, level, type, sync);
 
     const real rdx2 = 1. / (dx * dx);
     const real rdy2 = 1. / (dy * dy);
     const real rdz2 = 1. / (dz * dz);
 
     // preparation for diffusion step (alpha, beta)
-    real alphaX = rdx2;
-    real alphaY = rdy2;
-    real alphaZ = rdz2;
+    const real alphaX = rdx2;
+    const real alphaY = rdy2;
+    const real alphaZ = rdz2;
 
     const real rbeta = 2. * (alphaX + alphaY + alphaZ);
     const real beta = 1. / rbeta;
 
-#pragma acc data present(out, tmp)
+    const size_t max_it = m_diffusion_max_iter;
+    const real tol_res = m_diffusion_tol_res;
+
+    tmp.copy_data(out);
+#pragma acc data present(out, tmp, b)
     {
-        // initialization
-        // inner
-#pragma acc kernels present(d_iList[start_i:(end_i-start_i)]) async
-#pragma acc loop independent
-        for (size_t j = start_i; j < end_i; ++j) {
-            const size_t i = d_iList[j];
-            tmp[i] = out[i];
-        }
+        size_t it = 0;
+        real sum;
+        real res = 10000.;
 
-        // boundary
-#pragma acc kernels present(d_bList[start_b:(end_b-start_b)]) async
-#pragma acc loop independent
-        for (size_t j = start_b; j < end_b; ++j) {
-            const size_t i = d_bList[j];
-            tmp[i] = out[i];
-        }
+        const size_t neighbour_i = 1;
+        const size_t neighbour_j = Nx;
+        const size_t neighbour_k = Nx * Ny;
+        while (res > tol_res && it < max_it) {
+            ColoredGaussSeidelDiffuse::colored_gauss_seidel_step(out, b, alphaX, alphaY, alphaZ, beta, m_dsign, m_w, sync);
+            boundary->apply_boundary(data_out, level, type, sync); // for res/err only Dirichlet BC
```

Comment:
two spaces between code and comment are needed.

Reply:
b167247